test/parallel/test-fs-watch-recursive.js

```diff
@@ -20,32 +18,39 @@ const testsubdir = fs.mkdtempSync(testDir + path.sep);
 const relativePathOne = path.join(path.basename(testsubdir), filenameOne);
 const filepathOne = path.join(testsubdir, filenameOne);
 
-const watcher = fs.watch(testDir, { recursive: true });
+try {
+  const watcher = fs.watch(testDir, { recursive: true });
 
-let watcherClosed = false;
-watcher.on('change', function(event, filename) {
-  assert.ok(event === 'change' || event === 'rename');
+  let watcherClosed = false;
+  watcher.on('change', function(event, filename) {
+    assert.ok(event === 'change' || event === 'rename');
 
-  // Ignore stale events generated by mkdir and other tests
-  if (filename !== relativePathOne)
-    return;
+    // Ignore stale events generated by mkdir and other tests
+    if (filename !== relativePathOne)
+      return;
 
+    if (common.isOSX) {
+      clearInterval(interval);
+    }
+    watcher.close();
+    watcherClosed = true;
+  });
+
+  let interval;
   if (common.isOSX) {
-    clearInterval(interval);
+    interval = setInterval(function() {
+      fs.writeFileSync(filepathOne, 'world');
+    }, 10);
+  } else {
+    fs.writeFileSync(filepathOne, 'world');
   }
-  watcher.close();
-  watcherClosed = true;
-});
 
-let interval;
-if (common.isOSX) {
-  interval = setInterval(function() {
-    fs.writeFileSync(filepathOne, 'world');
-  }, 10);
-} else {
-  fs.writeFileSync(filepathOne, 'world');
+  process.on('exit', function() {
+    assert(watcherClosed, 'watcher Object was not closed');
+  });
+} catch (err) {
+  if (common.isOSX || common.isWindows)
+    throw err;
+  else if (err.code !== 'ERR_FEATURE_UNAVAILABLE_ON_PLATFORM')
+    throw err;
 }
```

Comment:
The try block is arguably way too big - it should really only be around the `fs.watch()` call - but it'd probably be even better to do just this at the top:
```js
if (!common.isOSX && !common.isWindows) {
  common.expectsErrror(() => fs.watch(testDir, { recursive: true }),
                       { code: 'ERR_FEATURE_UNAVAILABLE_ON_PLATFORM' });
}
```

Reply:
Thanks for the review, I'll edit the test soon. I believe the try should be around the whole test, as in the future one might try to add more logic regarding the recursive watch. Though I do acknowledge that there is no a great difference between the two alternatives currently.

Reply:
I tried @bnoordhuis first proposal with expectsErrror (of course I fix the typo to expectsError). Is it possible that the error just goes through? I mean I tested with Ubuntu, and it didn't caught the error (of course I put the logic inside an anonymous function so it wouldn't be called twice. The error was just thrown. 
Regarding the second proposal which @BridgeAR and @BridgeAR proposed, I think it's too problematic. The test,  attaches to process's exit event (see: "watcher Object was not closed"). As such a try cannot wrap only the fs.watch . if fs.watch is not supported, the whole logic should not continue.

Reply:
okay I used expectsError in the wrong way. I will send a fixed test soon.